File: 2577-minimum-time-to-visit-a-cell-in-a-grid/2577-minimum-time-to-visit-a-cell-in-a-grid.py
import heapq

class Solution:
    def minimumTime(self, grid):
        if grid[0][1] > 1 and grid[1][0] > 1:
            return -1

        n, m = len(grid), len(grid[0])
        dx = [0, 0, 1, -1]
        dy = [1, -1, 0, 0]
        visited = [[False] * m for _ in range(n)]
        pq = [] 
        heapq.heappush(pq, (0, 0, 0))

        while pq:
            cur_time, x, y = heapq.heappop(pq)
            if x == n - 1 and y == m - 1:
                return cur_time
            if visited[x][y]:
                continue
            visited[x][y] = True

            for i in range(4):
                nx, ny = x + dx[i], y + dy[i]
                if 0 <= nx < n and 0 <= ny < m:
                    wasted_time = 0 if (grid[nx][ny] - cur_time) % 2 else 1
                    max_time = max(cur_time + 1, grid[nx][ny] + wasted_time)
                    heapq.heappush(pq, (max_time, nx, ny))

        return -1

File: 2577-minimum-time-to-visit-a-cell-in-a-grid/2577-minimum-time-to-visit-a-cell-in-a-grid.java
import java.util.*;

class Solution {
    public int minimumTime(int[][] grid) {
        if (grid[0][1] > 1 && grid[1][0] > 1) {
            return -1;
        }

        int n = grid.length, m = grid[0].length;
        int[] dx = {0, 0, 1, -1};
        int[] dy = {1, -1, 0, 0};
        boolean[][] visited = new boolean[n][m];
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0])); 
        pq.offer(new int[]{0, 0, 0});

        while (!pq.isEmpty()) {
            int[] top = pq.poll();
            int curTime = top[0], x = top[1], y = top[2];

            if (x == n - 1 && y == m - 1) {
                return curTime;
            }
            if (visited[x][y]) {
                continue;
            }
            visited[x][y] = true;

            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i], ny = y + dy[i];
                if (nx >= 0 && ny >= 0 && nx < n && ny < m) {
                    int wastedTime = (grid[nx][ny] - curTime) % 2 == 0 ? 1 : 0;
                    int maxTime = Math.max(curTime + 1, grid[nx][ny] + wastedTime);
                    pq.offer(new int[]{maxTime, nx, ny});
                }
            }
        }

        return -1;
    }
}
